src/tools/pkcs11-tool.c

```diff
@@ -1648,6 +1648,37 @@ static void list_slots(int tokens, int refresh, int print)
 	}
 }
 
+static const char *
+get_uri(CK_TOKEN_INFO_PTR info)
+{
+	static char URI[1024];
+	unsigned int shift = 0;
+	strcat(URI, "pkcs11:");
+	shift += strlen("pkcs11:");
+	strcat(URI + shift, "model=");
+	shift += strlen("model=");
+	const char *model = percent_encode(info->model, sizeof(info->model));
+	strcat(URI + shift, model);
+	shift += strlen(model);
+	strcat(URI + shift, ";manufacturer=");
+	shift += strlen(";manufacturer=");
+	const char *manufacturer = percent_encode(info->manufacturerID, sizeof(info->manufacturerID));
+	strcat(URI + shift, manufacturer);
+	shift += strlen(manufacturer);
+	strcat(URI + shift, ";serial=");
+	shift += strlen(";serial=");
+	const char *serial = percent_encode(info->serialNumber, sizeof(info->serialNumber));
+	strcat(URI + shift, serial);
+	shift += strlen(serial);
+	strcat(URI + shift, ";token=");
+	shift += strlen(";token=");
+	const char *token = percent_encode(info->label, sizeof(info->label));
+	strcat(URI + shift, token);
+	shift += strlen(token);
```

Comment:
This already starts to feel like a lot of repetition (and it s not even checking the bounds of `URI` buffer (use also lowercase for the variables please).

I would prefer to use helper function for pushing a key+value to the buffer to avoid too much repetition and properly check buffer bounds.

Also, please do not use strcat. Its very easy to get it wrong if the buffers are not allocated large enough and some static analyzers are already allergic to the use of this function (coverity?).